contracts/CDPLeverage.sol

```diff
@@ -0,0 +1,456 @@
+*/
+contract CDPLeverage is DSMath {
+
+    // Constants and modules set at contract inception
+    ITub public tub;
+    DSToken public weth;
+    DSToken public peth;
+    DSToken public dai;
+    DSToken public gov;
+    DSValue public pip;
+    DSValue public pep;
+    uint256 public makerLR;
+    uint256 public fee;
+    uint256 public axe;
+    IMatchingMarket public dex;
+    uint constant WAD = 10 ** 18;
+    uint constant RAY = 10 ** 27;
+
+    // This Struct tracks position information for a specific investor address
+    struct Investor {
+        uint256 layers;          // Number of CDP layers. According to github schematic
+        uint256 principal;       // WAD - Principal contribution in Eth
+        uint256 collatRatio;     // WAD - Collaterazation ratio
+        bytes32 cdpID;           // cdpID
+        uint256 purchPrice;      // WAD; currPrice at time of purchase
+        uint256 daiAmountFinal;  // WAD - Amount of Dai that an investor has after leveraging
+        uint256 totalDebt;       // WAD - Total amount of debt (DAI)
+        uint256 priceFloor;      // WAD - Price floor
+        uint256 index;           // index of investor
+    }
+    mapping (address => Investor) public investors;
+    address[] public investorAddresses;
+
+
+    // Modifiers
+    modifier onlyInvestor {
+        require(investorAddresses.length != 0, "Investor Array is empty");
+        require(investorAddresses[investors[msg.sender].index] == msg.sender, "Investor has not opened a position yet");
+        _;
+    }
+
+
+    // Events
+    event OpenPosition(address owner, uint256 principalEth, uint256 purchPrice, uint256 layers);
+    event ClosePosition(address owner, uint256 sellPrice);
+
+
+    /**
+    * @dev Constructor function
+    * Collects module information from the tub and matchingMarket (OasisDex)contracts
+    */
+    constructor()
+        public
+    {
+        address _tub = 0xa71937147b55Deb8a530C7229C442Fd3F31b7db2;         //Kovan
+        address _oasisDex = 0x8cf1Cab422A0b6b554077A361f8419cDf122a9F9;    //Kovan
+        tub = ITub(_tub);
+        weth = tub.gem();
+        peth = tub.skr();
+        dai = tub.sai();
+        pip = tub.pip();
+        pep = tub.pep();
+        gov = tub.gov();
+        fee = tub.fee();
+        axe = tub.axe();
+        dex = IMatchingMarket(_oasisDex);
+
+        makerLR = ray2wad(tub.mat());
+
+        // we approve tub, tap, and dex to access weth, peth, and dai contracts
+        weth.approve(tub, 1000000000000000000000000);
+        weth.approve(dex, 1000000000000000000000000);
+        peth.approve(tub, 1000000000000000000000000);
+        dai.approve(tub, 10000000000000000000000000000);
+        dai.approve(dex, 10000000000000000000000000000);
+        gov.approve(tub, 1000000000000000000000000);
+        gov.approve(dex, 1000000000000000000000000);
+    }
+
+
+    // Fallback function
+    function() payable public {}
```

Comment:
Do we need this?

Reply:
Possibly. Was previously getting errors with ETH transfers from WETH contract to CDPLeverage contract.

Reply:
Yep, you're right. I forgot about W-ETH unwrapping